### ISSUE TITLE
Ignore directory if file is present

### ISSUE THREAD

**Issue by idavydov** [ncw/rclone#1808]
rdiff-backup and duplicity has an option `--exclude-if-present`. 

It acts as following: during sync, before proceeding with a directory, it checks if a file with specific name exists in it. If it does, the directory is excluded from sync.

My current workaround in bash looks like this:
```bash
mkexcl () {
    exclf=$(mktemp)
    (cd $1
     find -xdev -name .no-sync |
	 sed 's/^\.//;s/\.no-sync$/**/'
    ) > $exclf 2>/dev/null
    echo $exclf
 }

exclf=$(mkexcl $source)
rclone sync --exclude-from $exclf $source $destanation
```

but this is rather ugly and not very general.

Is it possible to add this feature to rclone?

**Comment by ncw** [comment]
I'm all for importing good ideas from elsewhere :-)

Is that the typical use for it that you add a `.no-sync` file?

Here is what the duplicity docs say

> --exclude-if-present filename
> Exclude directories if filename is present. Allows the user to specify folders that they do not wish to backup by adding a specified file (e.g. ".nobackup") instead of maintaining a comprehensive exclude/include list. This option needs to come before any other include or exclude options.

And rdiff-backup says essentially the same thing

       --exclude-if-present filename
              Exclude directories if filename is present. This option needs to
              come before any other include or exclude options.

This would be relatively easy to implement - fancy having a go?

**Comment by idavydov** [comment]
The typical use it to add a file, I guess there is no convention on the filename (if that is your question).

I will try to see what can be done. Which layer would you suggest updating? Adding a method to `Directory` interface & updating `walk.go` (just briefly looked at the source)?

What about the option name? `--exclude-if-present` to make it uniform?

**Comment by idavydov** [comment]
I have created a simple implementation.

The main problem is that it will first get the full list of files, which might be rather slow on directories with large number of files. The way to fix this, is to update the `Fs` interface, to have something like `Exists`, but this looks like a rather radical change.

**Comment by ncw** [comment]
> I have created a simple implementation.

And very good it is too :-)  Keeping the --exclude-if-present is perfect.

> The main problem is that it will first get the full list of files, which might be rather slow on directories with large number of files. The way to fix this, is to update the Fs interface, to have something like Exists, but this looks like a rather radical change.

Most of the Fs would have to list the directory anyway so I'd say this is a minor point.

**Comment by idavydov** [comment]
The feature was implemented (#1812), so we can close the issue.
